### PR TITLE
Use eslint v9 flat config

### LINKED ISSUE #465: Use eslint 9
With the introduction of turborepo and big updates, I suggest updating eslint to version 9 as well.
If it's okay, I'd like to work on it.

### PR DESCRIPTION
**Pull request by plbstl**
## Priority*

- [x] High: This PR needs to be merged first for other tasks.
- [ ] Middle: This PR should be merged quickly to prevent conflicts due to common changes. (default)
- [ ] Low: This PR does not affect other tasks, so it can be merged later.

## Purpose of the PR*

Use eslint version 9 flat config.

## Changes*

This PR upgrades `eslint` to version 9 and uses the flat config.

## How to check the feature

Add code that warns or errors when `eslint` is set up correctly.

## Reference

- Closes #465

### CODE REVIEW

File: eslint.config.mjs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  js.configs.recommended,
+  ...ts.configs.recommended,
+  reactPlugin.configs.flat.recommended,
+  jsxA11y.flatConfigs.recommended,
+  eslintPluginPrettierRecommended,
+
+  // Add compatibility
+  // @ts-expect-error - mismatched typings
+  ...fixupConfigRules(new FlatCompat().extends('plugin:react-hooks/recommended')),
+  // ...fixupConfigRules(new FlatCompat().extends('plugin:react-hooks/recommended', 'plugin:import/recommended')),
```
Comment (plbstl) [inline eslint.config.mjs:31]: There is one issue I'm facing, I cannot get `eslint-plugin-import` to work properly

Comment (PatrykKuniczak) [reply eslint.config.mjs:31]: It needs to work, because it's crucial 😄

File: eslint.config.mjs
Comment (PatrykKuniczak) [inline eslint.config.mjs:1]: Are you able to do it with `.mts`?
Because we want to have 100% `.ts` coverage in the near future

**Comment by PatrykKuniczak** [comment]
Let's remove also:
```ts
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
```

from `packages/i18n` and 
```ts
// eslint-disable-next-line
// @ts-ignore
```
from `pages/content-runtime/lib`

and check all `// eslint-disable-next-line` e.g. for `refresh` and `reload` because the most of it isn't necessary probably 😄

File: eslint.config.mjs
Comment (plbstl) [reply eslint.config.mjs:1]: This is currently an experimental feature. https://eslint.org/blog/2024/08/eslint-v9.9.0-released/#experimental-typescript-configuration-files. It may be stable in the near future

**Comment by plbstl** [comment]
Okay, I'll make those changes now

File: eslint.config.mjs
Comment (PatrykKuniczak) [reply eslint.config.mjs:1]: @plbstl If it works, go ahead :)

**Comment by PatrykKuniczak** [comment]
@plbstl Let's rebase ASAP, you'll avoid more conflicts, because i'm working on other new things 😄

File: eslint.config.mjs
Comment (plbstl) [reply eslint.config.mjs:1]: This works but for the CLI only, the vscode editor extension does not pick it up. I opened https://github.com/microsoft/vscode-eslint/issues/1927

**Comment by plbstl** [comment]
For the recent changes in #697, I was thinking `pnpm exec` may be better than `pnpx` (pnpm dlx) since it uses the locally installed packages.

https://pnpm.io/cli/exec#examples

File: package.json [hunk 2]
```diff
@@ -8,7 +8,7 @@
     "url": "https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite.git"
   },
   "scripts": {
-    "clean:node_modules": "pnpx rimraf --glob **/node_modules",
+    "clean:node_modules": "rimraf --glob **/node_modules",
```
Comment (PatrykKuniczak) [inline package.json:11]: This `pnpx` is here with reason, read description:
https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/pull/697#issue-2502937642

**Review by PatrykKuniczak — CHANGES_REQUESTED** [review]

**Comment by PatrykKuniczak** [comment]
@plbstl Are you able to set this `--flag unstable_ts_config` inside config file, instead of the flag in each script?

**Comment by plbstl** [comment]
I checked the source code, the option is only available as a CLI flag

File: eslint.config.mjs
Comment (plbstl) [reply eslint.config.mjs:1]: It works when this is added to settings.json

```json
 "eslint.options": {
    "flags": ["unstable_ts_config"]
  }
```

Comment (PatrykKuniczak) [reply eslint.config.mjs:1]: > It works when this is added to settings.json
> 
> ```json
>  "eslint.options": {
>     "flags": ["unstable_ts_config"]
>   }
> ```

Good, but it need to work for all IDE's :)

File: eslint.config.mjs (hunk 1 above)
Comment (plbstl) [reply eslint.config.mjs]: The latest change includes a fix for this. 
Also noticed that the `eslint-plugin-import` configuration in the legacy config only worked with javascript files.

**Comment by plbstl** [comment]
Replaced [`eslint-plugin-import`](https://github.com/import-js/eslint-plugin-import) with [`eslint-plugin-import-x`](https://github.com/un-ts/eslint-plugin-import-x) as it supports eslint v9.

It also has an additional rule [`no-rename-default`](https://github.com/un-ts/eslint-plugin-import-x/blob/master/docs/rules/no-rename-default.md) under [Helpful warnings](https://github.com/un-ts/eslint-plugin-import-x#helpful-warnings) section.

File: eslint.config.mjs
Comment (plbstl) [reply eslint.config.mjs:1]: Maybe a section to tell what needs to be done since that would be on a per editor basis. https://eslint.org/docs/latest/use/integrations#editors

For sublime: the CLI flag is inputted in the `args` eslint setting. See https://github.com/SublimeLinter/SublimeLinter-eslint#examples, https://www.sublimelinter.com/en/master/linter_settings.html#args.
Same with JetBrains IDEs. See step 5 of https://www.jetbrains.com/help/idea/eslint.html#ws_js_eslint_manual_configuration

Basically passing in `--flag unstable_ts_config` when loading eslint would suffice for most.

File: package.json (hunk 2 above)
Comment (plbstl) [reply package.json]: Okay

File: eslint.config.mjs
Comment (PatrykKuniczak) [reply eslint.config.mjs:1]: @plbstl In that case, leave this flag, it's the most certain solution

**Comment by PatrykKuniczak** [comment]
@plbstl You've merge main wrongly, because i see unnecessary files(which shouldn't be shown as change here) in `Files changed`

And you have a lot of conflicts, let's solve it

**Comment by plbstl** [comment]
@PatrykKuniczak The conflicts have now been resolved

**Comment by PatrykKuniczak** [comment]
It doesn't work at all
image

First of all, eslint config wasn't found be IDE
Second thing, when i've added 
tsconfig for this eslint:
```json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Chrome Extension Utils",
  "extends": "./packages/tsconfig/base.json",
  "compilerOptions": {
    "noEmit": false
  },
  "include": ["eslint.config.ts"]
}
```
and "ready": "tsc -p eslint.tsconfig.json && turbo ready" for `package.json` and i've run it.
But it still doesn't work because of the lack of types.

And after all we have this warns, which are the problem, why we not decide to bump `eslint` to version 9:
image

I think this PR is overkill for that simply solution as eslint is.
IDN what to do with it, maybe @Jonghakseo have sth to say about it.
IMO the best idea is wait a little, when this libs start to support eslint 9

**Comment by plbstl** [comment]
Can you tell me the name of the IDE? The repo does not contain any config files for IDEs. You have to instruct the specific IDE eslint plugin to load the config. Just as discussed in https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/pull/699?notification_referrer_id=NT_kwDOAuvH57QxMjIzNjkxMTQ2MTo0OTAwNjU2Nw#discussion_r1747710475 

`@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` are no longer needed, a mistake on my part for not removing the dependencies. See https://typescript-eslint.io/packages/typescript-eslint/#migrating-from-legacy-config-setups

Same goes for `eslint-plugin-jsx-a11y`. Updating `eslint-plugin-jsx-a11y` to `6.10` fixes the peer dependency issue.

I don't know where the `eslint-plugin-import` issue is coming from. `eslint-import-resolver-typescript` has correct peer dependencies (https://github.com/import-js/eslint-import-resolver-typescript/blob/01902942c42292ab3cd266c2d6855c76a2ab5d7e/package.json#L64) and `eslint-plugin-import-x` is used and it is not depending on `eslint-plugin-import`.

Since `eslint-config-airbnb-typescript` is archived (https://github.com/iamturns/eslint-config-airbnb-typescript), I don't think it'll ever support version 9 or even update its peer dependencies.

**Comment by PatrykKuniczak** [comment]
> Can you tell me the name of the IDE? The repo does not contain any config files for IDEs. You have to instruct the specific IDE eslint plugin to load the config. Just as discussed in #699
> 
> `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` are no longer needed, a mistake on my part for not removing the dependencies. See https://typescript-eslint.io/packages/typescript-eslint/#migrating-from-legacy-config-setups
> 
> Same goes for `eslint-plugin-jsx-a11y`. Updating `eslint-plugin-jsx-a11y` to `6.10` fixes the peer dependency issue.
> 
> I don't know where the `eslint-plugin-import` issue is coming from. `eslint-import-resolver-typescript` has correct peer dependencies (https://github.com/import-js/eslint-import-resolver-typescript/blob/01902942c42292ab3cd266c2d6855c76a2ab5d7e/package.json#L64) and `eslint-plugin-import-x` is used and it is not depending on `eslint-plugin-import`.
> 
> Since `eslint-config-airbnb-typescript` is archived (https://github.com/iamturns/eslint-config-airbnb-typescript), I don't think it'll ever support version 9 or even update its peer dependencies.

You've pinned This PR in `Just as discussed in https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/pull/699` it's correct?

I have been using WebStorm

**Comment by plbstl** [comment]
> You've pinned This PR in `Just as discussed in https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/pull/699` it's correct?

This is what I meant to link to: `Just as discussed in https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/pull/699#discussion_r1747710475`

**Comment by PatrykKuniczak** [comment]
@plbstl In the future, you can `Copy link`
image

It's easier to navigate by everybody, because it creates clickable link 😄

**Comment by PatrykKuniczak** [comment]
@plbstl Let's rebase branch

**Comment by plbstl** [comment]
> @plbstl In the future, you can `Copy link` image
> 
> It's easier to navigate by everybody, because it creates clickable link 😄

Thank you, that's what I did but for some reason it added a search param for a `notificationId` before the URL hash

**Review by PatrykKuniczak — COMMENTED** [review]
All your imports:
`import * as anything from "anything"`

should be:
`import anything from "anything"`

If it doesn't work you don't `include` or set `types` in tsconfig properly

**Comment by plbstl** [comment]
> All your imports: `import * as anything from "anything"`
> 
> should be: `import anything from "anything"`
> 
> If it doesn't work you don't `include` or set `types` in tsconfig properly

It works. Its just that eslint shows a warning for the [`import-x/no-named-as-default-member`](https://github.com/un-ts/eslint-plugin-import-x/blob/docs/rules/no-named-as-default-member.md) rule.

Should I change it back and turn off the rule? Or change the imports to named imports? Or something else?

**Comment by PatrykKuniczak** [comment]
> > All your imports: `import * as anything from "anything"`
> > should be: `import anything from "anything"`
> > If it doesn't work you don't `include` or set `types` in tsconfig properly
> 
> It works. Its just that eslint shows a warning for the [`import-x/no-named-as-default-member`](https://github.com/un-ts/eslint-plugin-import-x/blob/docs/rules/no-named-as-default-member.md) rule.
> 
> Should I change it back and turn off the rule? Or change the imports to named imports? Or something else?

Maybe set it off, because we're using default imports

**Comment by PatrykKuniczak** [comment]
@plbstl You're right, because no default is exported from there :)
image

Leave it, in the way you've done it

But there's greater problem. my IDE still can't find config, because of `.ts` i think this should be `.js` because it works.

Or use `.ts` but build this file on `postinstall` and IDE will be using bundled version of this, let's check if it will be working 😄

**Comment by plbstl** [comment]
> But there's greater problem. my IDE still can't find config, because of `.ts` i think this should be `.js` because it works.
> 
> Or use `.ts` but build this file on `postinstall` and IDE will be using bundled version of this, let's check if it will be working 😄
> 
> If yes, then add `.js` to .gitignore and it will work perfect :)
> 
> PS: With second approach, we can have 100% ts coverage, maybe it's worth efford 😄

The latest changes have done exactly that, and it works well.

**Comment by PatrykKuniczak** [comment]
@plbstl I'll check it tomorrow morning 😸

File: eslint.config.ts
Comment (PatrykKuniczak) [inline eslint.config.ts:1]: I have this error:
image

For those imports:
image

**Review by PatrykKuniczak — CHANGES_REQUESTED** [review]
Remove all `--flag unstable_ts_config` because now we have `.js` and we don't need to take a risk with `unstable` things

File: package.json [hunk 3]
```diff
@@ -36,34 +37,40 @@
     "react-dom": "18.3.1"
   },
   "devDependencies": {
+    "@eslint/compat": "^1.1.1",
+    "@eslint/eslintrc": "^3.1.0",
+    "@eslint/js": "^9.10.0",
```
[the remark this answers is not in the document]
Comment (plbstl) [reply package.json:43]: It may be better to leave it the default way as its more predictable and how most projects are structured.
This may not count but installing a new package re-sorts the dependencies alphabetically.

File: package.json [hunk 4]
```diff
@@ -27,6 +27,7 @@
     "lint:fix": "turbo lint:fix --continue -- --fix --cache --cache-location node_modules/.cache/.eslintcache",
     "prettier": "turbo prettier --continue -- --cache --cache-location node_modules/.cache/.prettiercache",
     "prepare": "husky",
+    "postinstall": "esbuild eslint.config.ts --outfile=eslint.config.js && prettier eslint.config.js --write --ignore-path .prettierignore",
```
[the remark this answers is not in the document]
Comment (plbstl) [reply package.json]: The latest changes implement this. But the name of the script is `build:eslint`

File: eslint.config.ts
Comment (plbstl) [reply eslint.config.ts:1]: A fix has been implemented for this

File: package.json [hunk 5]
```diff
@@ -16,6 +16,7 @@
     "clean:install": "pnpm clean:node_modules && pnpm install --frozen-lockfile",
     "build": "pnpm clean:bundle && turbo ready && turbo build",
     "build:firefox": "pnpm clean:bundle && turbo ready && cross-env __FIREFOX__=true turbo build",
+    "build:eslint": "esbuild eslint.config.ts --outfile=eslint.config.js",
```
Comment (PatrykKuniczak) [inline package.json:19]: It this case, when you have `tsconfig` you don't really need `esbuild` you can build it via `tsc`
I will probably remove all `esbuild` from entire project to simplify it, because it's harder and harder to maintain.

Let me know

Comment (plbstl) [reply package.json]: Implemented this in the latest change

**Comment by PatrykKuniczak** [comment]
@plbstl Rebase and resolve conflicts

File: package.json [hunk 6]
```diff
@@ -16,6 +16,7 @@
     "clean:install": "pnpm clean:node_modules && pnpm install --frozen-lockfile",
     "build": "pnpm clean:bundle && turbo ready && turbo build",
     "build:firefox": "pnpm clean:bundle && turbo ready && cross-env __FIREFOX__=true turbo build",
+    "build:eslint": "tsc --build",
```
Comment (PatrykKuniczak) [inline package.json:19]:
```suggestion
    "build:eslint": "tsc -p eslint.json",
```

IMO this should looks like this, because `tsconfig.json` isn't necessary,
but include .ts config inside `eslint.json`

IMO 2 configs for that simple task is overkill, becasue you won't have other files in root, like we have in packages

**Comment by plbstl** [comment]
I have merged in the main branch, but I did not run into merge conflicts

File: package.json (hunk 6 above)
Comment (plbstl) [reply package.json:19]: It was initially only `tsconfig.json`, I then added the `@extension/tsconfig/eslint.json` with the latest changes to keep all tsconfigs in one place.

> IMO 2 configs for that simple task is overkill, becasue you won't have other files in root, like we have in packages

What do you suggest?

Comment (PatrykKuniczak) [reply package.json]: @plbstl Your current solution is good for now.
i'll refactor it a little, if i'll be able, on my branch.

Good work <3

**Comment by PatrykKuniczak** [comment]
@plbstl 
I have errors still:
image

Because this packages doesn't export default nothing:
image

Let's change it to:
import * as name from 'name';

It's working :)

**Comment by plbstl** [comment]
> @plbstl 
> 
> I have errors still:
> 
> image
> 
> 
> 
> Because this packages doesn't export default nothing:
> 
> image
> 
> 
> 
> Let's change it to:
> 
> `import * as name from 'name';`
> 
> 
> 
> It's working :)

I think it's because the root tsconfig was removed. It has to be added back since it includes configuration for eslint.config.ts

VSCode does not error for that, it made me totally forget about it

**Comment by PatrykKuniczak** [comment]
@plbstl Yeah, the tsconfig are using also for this, but anyway i'm thinking if it's good, but yeah, do it like you said, but extends `base.json`

**Comment by plbstl** [comment]
I have made the relevant changes

File: package.json (hunk 6 above)
Comment (PatrykKuniczak) [reply package.json:19]: It works the same without this flag, in other packages we doesn't use it.
image

It's really necessary?

Comment (plbstl) [reply package.json:19]: Not necessarily, for this use case. See https://www.typescriptlang.org/docs/handbook/project-references.html#caveats-for-project-references

**Comment by PatrykKuniczak** [comment]
@plbstl 
image
Why? xD

Let's rebase locally and see those conflicts, if there's no conflicts, then IDN xD

File: package.json (hunk 6 above)
Comment (PatrykKuniczak) [reply package.json:19]: @plbstl Ok, let's leave it here, it's seems nice option.
I'll use it for all `tsc` :)

**Review by PatrykKuniczak — APPROVED** [review]
LGTM

**Review by PatrykKuniczak — CHANGES_REQUESTED** [review]
image

You did sth wrong for sure, let's rebase it and i'll try it again :)

**Comment by plbstl** [comment]
> image
> 
> You did sth wrong for sure, let's rebase it and i'll try it again :)

- The `@ts-ignore` pragma was removed from `packages/i18n/index.ts` as per https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/pull/699#issuecomment-2328860784.
- The file that is being imported is generated after a `pnpm build` or `pnpm dev`.
- The Action runs `pnpm type-check && pnpm build`. type-check is run before the file is generated. 

It can be changed that build is run first. Or add `@ts-ignore` back.

**Comment by PatrykKuniczak** [comment]
OOOKKK, that's what i'm talking about with @Jonghakseo 
I'll open PR, to solve this

**Comment by PatrykKuniczak** [comment]
It'll be fixed in #740 
After this, everything should be OK.
Thanks again for your contribution <3

**Comment by PatrykKuniczak** [comment]
@plbstl If you'll eager to contribute a little more visit our [Discord](https://discord.com/invite/4ERQ6jgV9a) and i can explain you more 😄

**Comment by plbstl** [comment]
@PatrykKuniczak yeah sure. I just joined the server 🚀

**Review by PatrykKuniczak — CHANGES_REQUESTED** [review]
I've resolve conflicts, and add 1 missing types, but theres still some work:
As you can see, there's a bunch of warns:
image

Let's fix it, but:
https://github.com/iamturns/eslint-config-airbnb-typescript
Isn't longer supported, you need to find sth else, which supporting eslint v9
`eslint-config-airbnb-base` is also have last update 3 years ago
`eslint-plugin-import` is still as a peer deps
And there's missing types:
image

Let's try to find solution, because we don't want to merge it, becasue of potential problems with deps, we want to avoid it, that's why we don't decided to implement it for now, but i hope you're able to fix it.

PS: For reproduce that error try e.g. `pnpm add -D nodemon -w`

**Comment by PatrykKuniczak** [comment]
@plbstl Are you eager to go on?

**Comment by plbstl** [comment]
Apologies for the extended delay on this PR. I’m back now and ready to provide updates and address any pending feedback.

Here’s a quick update on the current status:

- The `eslint-config-airbnb` package is still in the process of supporting v9 (https://github.com/airbnb/javascript/issues/2961).
- An alternative to [`iamturns/eslint-config-airbnb-typescript`](https://github.com/iamturns/eslint-config-airbnb-typescript) is [`Kenneth-Sills/eslint-config-airbnb-typescript`](https://github.com/Kenneth-Sills/eslint-config-airbnb-typescript), but the `eslint-config-airbnb` issue above, is a blocker.
- Both `eslint-config-prettier` and `eslint-plugin-prettier` are currently listed as dependencies, but only `eslint-plugin-prettier` was/is used.

**Observing the config file, `eslint-config-airbnb` was not used in the old eslint config and is still not being used in the flat config.**

**Comment by PatrykKuniczak** [comment]
@plbstl If `eslint-config-airbnb` isn't necessary, let's remove it and you'll be a step ahead.

For the future, let's update branch with `rebase` instead of `merge` :)

**Comment by plbstl** [comment]
> @plbstl If `eslint-config-airbnb` isn't necessary, let's remove it and you'll be a step ahead.

Okay, I have removed `eslint-config-airbnb-typescript` along with `eslint-config-prettier`.

> For the future, let's update branch with `rebase` instead of `merge` :)

Duely noted, I have now updated the branch with `rebase` :)

---

There are no `pnpm install` warnings anymore, and the missing typedefs warnings in `eslint.config.ts` are also resolved when opened in VSCode. I currently don't have access to WebStorm to confirm.

File: package.json (hunk 6 above)
Comment (PatrykKuniczak) [inline package.json:19]:
```suggestion
    "build:eslint": "tsc -b",
```

To match other modules from `dev` branch :)

**Review by PatrykKuniczak — CHANGES_REQUESTED** [review]
I have warning:
image

Try to avoid it

**Comment by PatrykKuniczak** [comment]
@plbstl Also tests failed, let's take a closer look on that.

**Comment by plbstl** [comment]
ESLint should no longer display a warning for an unspecified React version. An update to a deeper dependency is causing the e2e tests to fail. This is useful to note in case future test failures occur without related code changes.

**Review by PatrykKuniczak — APPROVED** [review]
LGTM

Very good job ❤️

**Review by Jonghakseo — APPROVED** [review]
